### PR TITLE
feat(core): Add ability to "ref" and "unref" pending ops

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit adds an ability to "ref" or "unref" pending ops.

Up to this point Deno had a notion of "async ops" and "unref async ops";
the former keep event loop alive, while the latter do not block event loop
from finishing. It was not possible to change between op types after 
dispatching, one had to decide which type to use before dispatch.

Instead of storing ops in two separate "FuturesUnordered" collections,
now ops are stored in a single collection, with supplemental "HashSet"
storing ids of promises that were "unrefed".

Two APIs were added to "Deno.core":
- "Deno.core.refOp(promiseId)" which allows to mark promise id 
to be "refed" and keep event loop alive (the default behavior)
- "Deno.core.unrefOp(promiseId)" which allows to mark promise 
id as "unrefed" which won't block event loop from exiting

Closes https://github.com/denoland/deno/issues/12745

TODO:
- [x] metrics for unrefed ops (I think that unrefing ops should decrement `ops_dispatched_async` and increment `ops_dispatched_async_unref`, though it seems overly complicated)
- [x] cleanup handling of finished ops (remove `AsyncOpIterator`?)
- [x] add `id` field to promises returned from `Deno.core.opcallAsync`
- [x] add `Deno.core.ref(...promiseIds)` and `Deno.core.unref(...promiseIds)`

### CODE REVIEW

File: core/01_core.js
Comment (bartlomieju) [inline core/01_core.js:141]: This change was done so that ops that are unref by default don't have to call into Rust immediately after dispatch to mark them as "unrefed"

File: core/runtime.rs
Comment (bartlomieju) [inline core/runtime.rs:1547]: Clean this up

File: core/01_core.js
Comment (piscisaureus) [inline core/01_core.js:146]: Have you considered using a symbol here or a reverse WeakMap?
(I am not sure what that would do for performance, we'd have to measure that)

File: core/lib.deno_core.d.ts
Comment (piscisaureus) [inline core/lib.deno_core.d.ts:27]: Why does it take multiple `promiseId`s?
I do understand that it is faster to coalesce multiple `refOp` calls into a single call, but I wonder how common that is, and also note that passing a single int into the binding layer is faster than passing an array.

File: core/01_core.js
Comment (bartlomieju) [reply core/01_core.js:146]: I did not, but using a symbol might be preferable, that way we can better communicate that users shouldn't use that value. I'll explore that.

File: core/runtime.rs
Comment (piscisaureus) [inline core/runtime.rs:158]: 👍

**Review by piscisaureus — COMMENTED** [review]
Looks pretty good, some questions though.

**Comment by AaronO** [comment]
Do we want all op(calls) to be unrefable ? I feel like it should be opt-in and take its distinct path (e.g: `opUnref` instead of overloading `opAsync`)

File: core/lib.deno_core.d.ts
Comment (bartlomieju) [reply core/lib.deno_core.d.ts:27]: In some resources we might have multiple pending ops associated with it. So instead of crossing boundary multiple times I thought it's better to do it in one go. Keep in mind that this is spread operator so you're passing multiple integers and not an array (`Deno.core.refOps(1, 2, 3, 4)`). Also is there a situation where this API would be used in a hot path?

**Comment by piscisaureus** [comment]
> Do we want all op(calls) to be unrefable ? I feel like it should be opt-in and take its distinct path (e.g: `opUnref` instead of overloading `opAsync`)

I don't see a reason why we would be more restrictive about this than necessary.

**Comment by AaronO** [comment]
> > Do we want all op(calls) to be unrefable ? I feel like it should be opt-in and take its distinct path (e.g: `opUnref` instead of overloading `opAsync`)
> 
> I don't see a reason why we would be more restrictive about this than necessary.

Well if in practice there's only a very small subset of ops/opcalls that may be unref'd then I think it's cleaner to have its own distinct implementation path so as not to overload the simple async opcall path

File: core/lib.deno_core.d.ts
Comment (piscisaureus) [reply core/lib.deno_core.d.ts:27]: > Keep in mind that this is spread operator so you're passing multiple integers and not an array (Deno.core.refOps(1, 2, 3, 4))

I see, in the binding layer you're just iterating over `FunctionCallbackArguments`, that's pretty efficient and I have no concerns about it.

> In some resources we might have multiple pending ops associated with it.

You'll find that this almost never happens in practice but oh well.

Comment (bartlomieju) [reply core/lib.deno_core.d.ts:27]: If you prefer to have it accept single ID that is fine for me as well. Your call

File: core/01_core.js
Comment (piscisaureus) [reply core/01_core.js:141]: I probably would keep the old signature. The only "unref-by-default" ops are signal listeners and they're not a performance concern.

Comment (bartlomieju) [reply core/01_core.js:141]: Reverted.

File: core/01_core.js
Comment (bartlomieju) [reply core/01_core.js:146]: Used `Symbol.for("Deno.core.internalPromiseId")` instead, let's see if performance is better or worse

**Comment by bartlomieju** [comment]
> > > Do we want all op(calls) to be unrefable ? I feel like it should be opt-in and take its distinct path (e.g: `opUnref` instead of overloading `opAsync`)
> > 
> > 
> > I don't see a reason why we would be more restrictive about this than necessary.
> 
> Well if in practice there's only a very small subset of ops/opcalls that may be unref'd then I think it's cleaner to have its own distinct implementation path so as not to overload the simple async opcall path

I agree with Bert, I also believe it makes it more confusing to have distinct type of op for that purpose. Besides having different API there would be no upside IMO as we won't be able to validate if you are unrefing an op of that type. I'll be happy to revisit the topic after we add `.ref()`/`.unref()` to several APIs.

**Comment by AaronO** [comment]
> > > > Do we want all op(calls) to be unrefable ? I feel like it should be opt-in and take its distinct path (e.g: `opUnref` instead of overloading `opAsync`)
> > > 
> > > 
> > > I don't see a reason why we would be more restrictive about this than necessary.
> > 
> > 
> > Well if in practice there's only a very small subset of ops/opcalls that may be unref'd then I think it's cleaner to have its own distinct implementation path so as not to overload the simple async opcall path
> 
> I agree with Bert, I also believe it makes it more confusing to have distinct type of op for that purpose. Besides having different API there would be no upside IMO as we won't be able to validate if you are unrefing an op of that type. I'll be happy to revisit the topic after we add `.ref()`/`.unref()` to several APIs.

I'm not proposing to have a different kind of op rust-side just that the opCall JS side is explicit (killing `op_async_unref` is good IMO).

> > Do we want all op(calls) to be unrefable ? I feel like it should be opt-in and take its distinct path (e.g: `opUnref` instead of overloading `opAsync`)
> 
> I don't see a reason why we would be more restrictive about this than necessary.

It all depends on our call expectations but IMO:
1. `Deno.core.opAsync("op_foo", A, B, unref=true/false)` makes most sense if `unref` is derived from end-user args/inputs
2. `Deno.core.opUnref("op_foo", A, B)` makes more sense if the implementor decides if the opcall is unrefable 

From my understanding `unref=...` shouldn't really be derived from user inputs, it will mostly be hardcoded by the implementor (`ext/*/*.js` code)

If that's the case I think `Deno.core.opUnref` is the better choice, it explicits the intention and is easier to search/scan.

I also think it's more ergonomic, the previous `opAsync()` sig maps quite naturally to the function sig, one could you view it as a curried version of the opcall or `Deno.core.ops.foo_bar(A, B)` (which would be the API if we did Fast API), tagging an op-layer control value to the end of the args feels out of place

**Comment by bartlomieju** [comment]
> I'm not proposing a different kind of op rust-side just that the opCall JS side is explicit (killing op_async_unref is good IMO).

Okay, I misunderstood then. @AaronO I removed `unref` argument in `Deno.core.opAsync`; now you need to manually call `Deno.core.unrefOps()` afterwards. Is that okay for you?

**Comment by bartlomieju** [comment]
Also I'd appreciate some feedback on how to tackle metrics - right now there are two uncalled APIs, this needs to be addressed before landing.

File: core/ops_metrics.rs
Comment (piscisaureus) [inline core/ops_metrics.rs:89]: Just delete this, I guess?
@AaronO any opinions?

File: core/01_core.js
Comment (piscisaureus) [inline core/01_core.js:137]: Unnecessary change now.

**Review by piscisaureus — APPROVED** [review]
LGTM, let's land this and see what it does to the benchmarks.

File: core/ops_metrics.rs
Comment (bartlomieju) [reply core/ops_metrics.rs:89]: I guess that's fine. Right now "unrefed" ops don't take part in `opSanitizer` but actually it might be preferable to include them.

**Review by AaronO — APPROVED** [review]
LGTM

File: core/01_core.js
Comment (bnoordhuis) [inline core/01_core.js:47]: Non-blocking but as a possible future improvement: you can use a `v8::Private` if you're worried about the symbol leaking out to user land code (it's visible to `Object.getOwnPropertySymbols()` after all.)

Private symbols are like regular symbols except more... er, private. (Great explanation, Ben. +1, would read again.)

File: core/bindings.rs
Comment (bnoordhuis) [inline core/bindings.rs:478]: Minor caveat: the way try_from() is currently implemented for v8::Integer means it only works for numbers in the range -2**31..2**32. I suppose that's fine (I see you're using i32 elsewhere) but I'd be remiss if I didn't point it out.

`Value::integer_value()` and `Integer::value()` return i64s so you could be forgiven for assuming try_from() works for all numbers that fit in an i64, but unfortunately it currently doesn't.

edit: opened https://github.com/denoland/rusty_v8/issues/839

File: core/ops_metrics.rs
Comment (bnoordhuis) [inline core/ops_metrics.rs:19]: What is the reason for not removing them? Are they considered part of a public or stable(ish) API?

File: core/runtime.rs
Comment (bnoordhuis) [inline core/runtime.rs:158]: Should that be i64? You're casting `v8::Integer::value()` to u64 but that can give surprising results for negative numbers.

(Not really surprising, of course, just probably not what you want.)

File: core/runtime.rs
Comment (bnoordhuis) [inline core/runtime.rs:788]: Suggestion: rename this to e.g. `has_pending_refed_ops` or something like that? Makes it a little clearer that the logic here is "if more pending ops than unref'd ops exist, we have ref'd ops."

File: core/01_core.js
Comment (bnoordhuis) [inline core/01_core.js:141]:
```suggestion
    // Save the id on the promise so it can later be ref'ed or unref'ed
```

**Review by bnoordhuis — APPROVED** [review]
Nice work!

File: core/01_core.js
Comment (bartlomieju) [reply core/01_core.js:47]: Is that construct available in JS land?

File: core/bindings.rs
Comment (bartlomieju) [reply core/bindings.rs:478]: I suggest to handle that in a follow up PR; we got a bunch of places using the same conversion.

File: core/ops_metrics.rs
Comment (bartlomieju) [reply core/ops_metrics.rs:19]: Yes, this are part of stable `Deno.Metrics` interface

File: core/runtime.rs
Comment (bartlomieju) [reply core/runtime.rs:158]: Sure, that makes more sense, maybe even `i32` as there are some other conversions going on in the code.

File: core/runtime.rs
Comment (bartlomieju) [reply core/runtime.rs:788]: Ok

File: core/01_core.js
Comment (bnoordhuis) [reply core/01_core.js:47]: Not yet, but I can add it. You may have seen node's getHiddenValue() and setHiddenValue() - it's the same mechanism.

Comment (bartlomieju) [reply core/01_core.js:47]: > Not yet, but I can add it. You may have seen node's getHiddenValue() and setHiddenValue() - it's the same mechanism.

I haven't seen it. I guess I'll add a TODO and we can change it once the API is added.

File: core/bindings.rs
Comment (andreubotella) [inline core/bindings.rs:511]: Can we check that `promise_id` is in `state.pending_ops`?

Comment (bartlomieju) [reply core/bindings.rs:511]: No, we'd have to iterate over all `pending_ops` killing the performance.

Comment (bartlomieju) [reply core/bindings.rs:511]: Oh wait, this is `bindings.rs`, it's not a hot path. I guess I can add a check here.

Comment (bartlomieju) [reply core/bindings.rs:511]: Actually not possible to do that, there's no way to inspect `promise_id` when iterating `pending_ops`

Comment (bartlomieju) [reply core/bindings.rs:511]: This might be possible once @AaronO moves promise ring to Rust (#12314)

Comment (andreubotella) [reply core/bindings.rs:511]: In the meantime, what probably can be done is change the `unrefOp` JS function to check whether the promise ID exists and hasn't resolved, right?

It would be pretty bad if some careless code added an already-resolved promise ID to `unrefed_ops`, since all that matters is the length compared to `pending_ops`. But I suppose CI would catch that, though.

Comment (bartlomieju) [reply core/bindings.rs:511]: > In the meantime, what probably can be done is change the unrefOp JS function to check whether the promise ID exists and hasn't resolved, right?

That would require to adding a wrapper around `Deno.core.unrefOp` in `core/01_core.js` just to verify the promise map.

I'm not sure if it's worth doing that at this moment; I think we should try this API in the wild and see if that's a problem.

**Pull request merged**
